### PR TITLE
PassThruContainer

### PR DESCRIPTION
**Pull request by asodeur**
This is an attempt at #3282. `PassThruContainer` is added to `numba.jitclass.passthru`. Usage looks like
```python
from numba import PassThruContainer

obj = dict(a=1)
container = PassThruContainer(obj)

@jit(nopython=True)
def identity(c):
    with objmode(a='int64'):
        c.obj['b'] = 2
        a = c.obj['a']

    return c, a

res, a = identity(container)

assert c is res
assert a == 1
assert res.obj['b'] == 2
```

The first version of  `PassThruContainer` was implemented as subclass of `jitclass._box.Box` but has since been rewritten to use `NRT_meminfo_new_from_pyobject` based on @sklam's comments below.

### CODE REVIEW

**Comment by sklam** [comment]
I think you can avoid `jitclass._box.Box` and `pyapi.object_set_private_data` and cut a lot of code if you use [NRT_meminfo_new_from_pyobject](https://github.com/numba/numba/blob/fdbe264706881d19e97494e5ca794c540479a4d0/numba/runtime/_nrt_python.c#L31). Both the numpy array and the unicode support uses it.  It will handle the clean-up of the PyObject references for you.

**Comment by sklam** [comment]
If you start with something like the unicode impl https://github.com/numba/numba/blob/fdbe264706881d19e97494e5ca794c540479a4d0/numba/unicode.py#L33-L44 and reduce the members just to the `meminfo`.  The unboxer can just set `meminfo` with `c.pyapi.nrt_meminfo_new_from_pyobject`.  The boxer can just return the python PassThru object associated with the meminfo.

**Comment by asodeur** [comment]
Using `NRT_meminfo_new_from_pyobject` and leaving definitions of any extra attributes to derived types (if required) did simplify things a lot.

We refactored our internal code to no longer require the creation of pass through types on the native side. Not sure if the [native constructor](https://github.com/numba/numba/blob/ed2f180d983a7682db6b44204e1cddb47fd5ecc5/numba/jitclass/passthru.py#L80) should stay. Since `NRT_meminfo_new_from_pyobject` is being creative with `MemInfo.data` the constructor has to write to `MemInfo.data` which does not sound like a good idea. (Would not need to touch `numba/runtime` at all w/o the native constructor)

Remaining open issues:
- [ ] Is there a use-case for the native constructor at all?
- [ ] passthru.py is implementing `operator.is_` for `Opaque` types. If this is considered generally useful
the implementaion should probably move to `targets\builtins.py`, otherwise it must be typed more
narrow
- [ ] Tests in test_passthru are not strictly isolated. Tests failing to call all `__del__`s might cause
later tests to fail
- [ ] Not sure about the `hash` implementation is a good one, seems to work for now though

**Comment by sklam** [comment]
Sorry for the delay.  

> Is there a use-case for the native constructor at all?

No, we should remove it.

>  With box.Box no longer used numba.jitclass is probably no longer the right place for passthru.py. Where shall it move?

Yes.  `numba/passthru.py` is fine.


I'll be looking more into the code to answer the other questions you have.

**Comment by asodeur** [comment]
Latest version has the native constructor removed, added some docs,  and has moved to `numba/passthru.py`. There is still some code that should probably move into other modules or be more narrowly typed:

1. `int(Opaque)` is implemented, could have been called `id(Opaque)` as well. Essentially this is `ptrtoint`. If this is considered generally useful for all subtypes of `Opaque` this should move to `targets/numbers.py` (or `targets/builtins.py` if we call it `id`). Otherwise, restrict to `passthru.opaque_pyobject`
2. `operator.is_(Opaque, Opaque)` is implemented as comparing the pointers for equality. Again, should move to `targets/builtins.py` or br restricted to `passthru.opaque_pyobject`
3. The implementation of `PassThruContainerType.__hash__` turned out to be essentially an implementation of `pyhash.c/_Py_HashPointer`. If considered useful, a generic implementation of `__hash__` for `Opaque` and subtypes could be extracted (should probably live in `targets/hashing.py`)

**Comment by asodeur** [comment]
After rewrite to `NRT_meminfo_new_from_pyobject` this does not strictly need to be part of core Numba and we have been using a version refactored as stand-alone package for some time now internally. A good reason to keep it with the core package (and especially the continous build) might be the issue from the [previous comment](https://github.com/numba/numba/pull/3640#issuecomment-467330259) which will likely collide with core Numba some day.

File: numba/passthru.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    val = cgutils.create_struct_proxy(typ)(
+        context.context, context.builder, value=val
+    )
+
+    return context.box(pass_thru_type, val.container)
+
+
+@lower_builtin(int, types.Opaque)
+def opaque_to_int(context, builder, sig, args):
+    return builder.ptrtoint(args[0], cgutils.intp_t)
```
Comment (eric-wieser) [inline numba/passthru.py:178]: Wouldn't this make more sense as an overload of `id`?

File: numba/passthru.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+    lhs_type, rhs_type = sig.args
+    # the lhs and rhs have the same type
+    if lhs_type == rhs_type:
+        lhs_ptr = builder.ptrtoint(args[0], cgutils.intp_t)
+        rhs_ptr = builder.ptrtoint(args[1], cgutils.intp_t)
+
+        return builder.icmp_unsigned('==', lhs_ptr, rhs_ptr)
+    else:
+        return cgutils.false_bit
```
Comment (eric-wieser) [inline numba/passthru.py:208]: Split this out to #5834, since its absense causes a bug.

File: numba/passthru.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        if y is PassThruContainerType():
+            def pass_thru_container_pass_thru_container_eq_impl(x, y):
+                return x.wrapped_obj is y.wrapped_obj
+
+            return pass_thru_container_pass_thru_container_eq_impl
+        else:
+            def pass_thru_container_any_eq_impl(x, y):
+                return False
+
+            return pass_thru_container_any_eq_impl
```
Comment (eric-wieser) [inline numba/passthru.py:223]: What about the case when  `y` is a `PassThruContainerType` but `x` is not?

Isn't the right thing to do here to return `None` and let someone else try in all but the case when both objects are `PassThruContainerType`?

File: numba/passthru.py (hunk 1 above)
Comment (asodeur) [reply numba/passthru.py:178]: Was not sure whether to name it `int` or `id` myself. Guess I'll rename to `id` then.

Comment (eric-wieser) [reply numba/passthru.py:178]: If you call it `id`, are the semantics of `id(pass_through_container.obj)` the same inside and outside of jitted code? If yes, then `id` is the right name.

Comment (eric-wieser) [reply numba/passthru.py:178]: This probably then belongs in `cpython/builtins.py`.

File: numba/passthru.py (hunk 3 above)
Comment (asodeur) [reply numba/passthru.py:223]: Your right, aside from being buggy this is definitely overreaching. Should only handle the case where `x` and `y` are `PassThruContainerType`

**Comment by eric-wieser** [comment]
I wonder if anything can be done to make `lower_constant(PassThruContainerType)` possible.

**Comment by asodeur** [comment]
I've not implemented `lower_constant` for `PassThruContainerType` before but for a number of other extension types. Essentially the strategy is to unbox the PyObject in the `lower_constant` implementation. However, there are a number of pitfalls and I have not found a generic solution for all of them:
1. Currently you are expected to return a borrowed reference from `lower_constant` which rules out re-using the unboxer of any mem-managed type (see #4494, there is #4506 but not ready for Numba>0.48, yet)
2. Types that are using the `EnvironmentManager` during (un)boxing cannot be unboxed in `lower_constant`
3. If caching should be possible you'll need some way to re-create the object (from `pickle` or global function in `objmode`)
4. It's not obvious how to clean-up the object. (This should not be an issue for a global variable but could escalate quickly if the constant was created from a closure) 
5. Clean-up is even harder b/c `lower_constant` might be called several times during compilation but the constants get fused by the linker

**Comment by luk-f-a** [comment]
@asodeur thanks creating this PR, this container can be very useful. Regarding `lower_constant`, I was wondering if it's really needed since the following works:

```python
from numba import jit, objmode

obj = dict(a=1)

@jit(nopython=True)
def identity():
    with objmode(a='int64'):
        obj['b'] = 2
        a = obj['a']

    return a

a = identity()

assert a == 1
assert obj['b'] == 2
```

ie, even if the container cannot be lowered as constant, the original object can, which should fulfill the use-case, right?

**Pull request closed without merging**